File: src/android/DBMeter.java
package org.apache.cordova.dbmeter;

import android.Manifest;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioRecord;
import android.media.AudioTrack;
import android.media.MediaRecorder;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.LOG;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Timer;
import java.util.TimerTask;

/**
 * This plugin provides the decibel level from the microphone.
 */
public class DBMeter extends CordovaPlugin {

    private static final String LOG_TAG = "DBMeter";
    private static final int REQ_CODE = 0;
    private AudioRecord audioRecord;
    private short[] buffer;
    private Timer timer;
    private boolean isListening = false;


    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        //LOG.setLogLevel(LOG.DEBUG);

        if (!PermissionHelper.hasPermission(this, Manifest.permission.RECORD_AUDIO)) {
            PermissionHelper.requestPermission(this, REQ_CODE, Manifest.permission.RECORD_AUDIO);
        } else {
            if (action.equals("start")) {
                this.start(callbackContext);
            } else if (action.equals("stop")) {
                this.stop(callbackContext);
            } else if (action.equals("isListening")) {
                this.isListening(callbackContext);
            } else if (action.equals("destroy")) {
                this.destroy(callbackContext);
            } else {
                LOG.e(LOG_TAG, "Not a valid action: " + action);
                return false;
            }
        }
        return true;
    }

    /**
     * Permits to free the memory from the audioRecord instance
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void destroy(CallbackContext callbackContext) {
        if (this.audioRecord != null) {
            this.isListening = false;
            this.audioRecord.stop();
            this.audioRecord = null;
            if (this.timer != null) {
                this.timer.cancel();
                this.timer = null;
            }
            callbackContext.success();
        } else {
            sendPluginError(callbackContext, PluginError.DBMETER_NOT_INITIALIZED, "DBMeter is not initialized");
        }
    }

    /**
     * Starts listening the audio signal and sends dB values as a
     * {@link org.apache.cordova.PluginResult PluginResult} using a
     * {@link java.util.TimerTask TimerTask}.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void start(final CallbackContext callbackContext) {
        final DBMeter that = this;
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                if (that.audioRecord == null) {
                    that.isListening = false;
                    int rate = AudioTrack.getNativeOutputSampleRate(AudioManager.STREAM_SYSTEM);
                    int bufferSize = AudioRecord.getMinBufferSize(rate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);

                    that.audioRecord = new AudioRecord(
                            MediaRecorder.AudioSource.MIC,
                            rate,
                            AudioFormat.CHANNEL_IN_MONO,
                            AudioFormat.ENCODING_PCM_16BIT,
                            bufferSize);

                    that.buffer = new short[bufferSize];
                }
                if (!that.isListening) {
                    that.isListening = true;
                    that.audioRecord.startRecording();

                    that.timer = new Timer(LOG_TAG, true);

                    //start calling run in a timertask
                    TimerTask timerTask = new TimerTask() {
                        public void run() {
                            int readSize = that.audioRecord.read(that.buffer, 0, that.buffer.length);
                            double db = 0;
                            double maxAmplitude = 0;
                            for (int i = 0; i < readSize; i++) {
                                if (Math.abs(that.buffer[i]) > maxAmplitude) {
                                    maxAmplitude = Math.abs(that.buffer[i]);
                                }
                            }

                            if (maxAmplitude != 0) {
                                db = 20.0 * Math.log10(maxAmplitude / 32767.0) + 90;
                            }

                            LOG.d(LOG_TAG, Double.toString(db));

                            PluginResult result = new PluginResult(PluginResult.Status.OK, (float) db);
                            result.setKeepCallback(true);
                            callbackContext.sendPluginResult(result);
                        }
                    };
                    that.timer.scheduleAtFixedRate(timerTask, 0, 100);
                }
            }
        });
    }

    /**
     * Stops listening the audio signal.
     * Even if stopped, the {@link android.media.AudioRecord AudioRecord} instance still exist.
     * To destroy this instance, please use the {@link #destroy(CallbackContext) destroy} method.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void stop(final CallbackContext callbackContext) {
        final DBMeter that = this;
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                if (that.isListening) {
                    that.isListening = false;


                    if (that.timer != null) {
                        that.timer.cancel();
                    }
                    if (that.audioRecord != null) {
                        that.audioRecord.stop();
                    }
                    callbackContext.success();

                } else {
                    sendPluginError(callbackContext, PluginError.DBMETER_NOT_LISTENING, "DBMeter is not listening");
                }
            }
        });
    }

    /**
     * Returns whether the DBMeter is listening.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void isListening(final CallbackContext callbackContext) {
        PluginResult result = new PluginResult(PluginResult.Status.OK, this.isListening);
        callbackContext.sendPluginResult(result);
    }

    /**
     * Convenient method to send plugin errors.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     * @param error           The error code to return
     * @param message         The error message to return
     */
    private void sendPluginError(CallbackContext callbackContext, PluginError error, String message) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("code", error.ordinal());
            jsonObject.put("message", message);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        callbackContext.error(jsonObject);
    }

    public enum PluginError {
        DBMETER_NOT_INITIALIZED,
        DBMETER_NOT_LISTENING
    }
}

File: plugin/src/ios/DBMeter.swift
import Foundation
import AVFoundation

@objc(DBMeter) class DBMeter : CDVPlugin {

    private let LOG_TAG = "DBMeter"
    private let REQ_CODE = 0
    private var isListening: Bool = false
    private var audioRecorder: AVAudioRecorder!
    private var command: CDVInvokedUrlCommand!
    private var timer: DispatchSourceTimer!
    private var isTimerExists: Bool = false

    /**
     This plugin provides the decibel level from the microphone.
     */
    init(commandDelegate: CDVCommandDelegate) {
        super.init()
        self.commandDelegate = commandDelegate
    }

    /**
     Permits to free the memory from the audioRecord instance
     */
    @objc(destroy:)
    func destroy(command: CDVInvokedUrlCommand) {
        if (self.isListening) {
            self.timer.suspend()
            self.isListening = false
        }

        self.command = nil

        if (self.audioRecorder != nil) {

            self.audioRecorder.stop()
            self.audioRecorder = nil

            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
            self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
        } else {
            self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_INITIALIZED, errorMessage: "DBMeter is not initialized")
        }
    }

    /**
     Starts listening the audio signal and sends dB values as a
     CDVPluginResult keeping the same calback alive.
     */
    @objc(start:)
    func start(command: CDVInvokedUrlCommand) {
        self.commandDelegate!.run(inBackground:{
            self.command = command

            if (!self.isTimerExists) {
                self.timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
                self.timer.scheduleRepeating(deadline: .now(), interval: DispatchTimeInterval.milliseconds(300))
                self.timer.setEventHandler(handler: self.timerCallBack)
                self.isTimerExists = true
            }

            if (self.audioRecorder == nil) {
                let url: URL = URL.init(fileURLWithPath: "/dev/null")

                let settings : [String:Any] = [
                    AVFormatIDKey: Int(kAudioFormatAppleLossless),
                    AVSampleRateKey: 44100.0,
                    AVNumberOfChannelsKey: 1 as NSNumber,
                    AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
                ]

                do {
                    let audioSession: AVAudioSession = AVAudioSession.sharedInstance()
                    try audioSession.setCategory(AVAudioSessionCategoryPlayAndRecord)
                    try audioSession.setActive(true)

                    self.audioRecorder = try AVAudioRecorder(url: url, settings: settings)
                    self.audioRecorder.isMeteringEnabled = true
                } catch {
                    self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_INITIALIZED, errorMessage: "Error while initializing DBMeter")
                }
            }
            if (!self.isListening) {
                self.isListening = true
                self.audioRecorder.record()

                self.timer.resume()
            }
        })
    }

    /**
     Stops listening the audio signal.
     Even if stopped, the AVAudioRecorder instance still exist.
     To destroy this instance, please use the destroy method.
     */
    @objc(stop:)
    public func stop(command: CDVInvokedUrlCommand) {
        self.commandDelegate!.run(inBackground: {
            if (self.isListening) {
                self.isListening = false

                if (self.audioRecorder != nil && self.audioRecorder.isRecording) {
                    self.timer.suspend()
                    self.audioRecorder.stop()
                }

                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            } else {
                self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_LISTENING, errorMessage: "DBMeter is not listening")
            }
        })
    }

    /**
     Returns whether the DBMeter is listening.
     */
    @objc(isListening:)
    func isListening(command: CDVInvokedUrlCommand?) -> Bool {
        if (command != nil) {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: self.isListening)
            self.commandDelegate!.send(pluginResult, callbackId: command!.callbackId)
        }
        return self.isListening;
    }

    private func timerCallBack() {
        autoreleasepool {
            if (self.isListening && self.audioRecorder != nil) {
                self.audioRecorder.updateMeters()

                let peakPowerForChannel = pow(10, (self.audioRecorder.averagePower(forChannel: 0) / 20))
                let db = Int32(round(20 * log10(peakPowerForChannel) + 90))

                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: db)
                pluginResult!.setKeepCallbackAs(true)
                self.commandDelegate!.send(pluginResult, callbackId: self.command.callbackId)
            }
        }
    }

    /**
     Convenient method to send plugin errors.
     */
    private func sendPluginError(callbackId: String, errorCode: PluginError, errorMessage: String) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: ["code": errorCode.hashValue, "message": errorMessage])
        self.commandDelegate!.send(pluginResult, callbackId: callbackId)
    }

    enum PluginError: String {
        case DBMETER_NOT_INITIALIZED = "0"
        case DBMETER_NOT_LISTENING = "1"
    }
}
